### PR TITLE
Add dynamic linq to EF MVC tutorial

### PR DESCRIPTION
**Pull request by tdykstra**
@divega This is code I adapted for the tutorial from an example provided by @BillWagner, in response to customer comments that hard-coding column names for filter expressions was impractical in some scenarios. What's new here is the `SortCollectionBy<T>` method.

[Review URL](https://review.docs.microsoft.com/en-us/aspnet/core/data/ef-mvc/advanced?branch=td-dynamiclinq#dynamic-linq)

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@bricelam can you review this or suggest an EF asset?

**Comment by bricelam** [comment]
@anpete is your subject matter expert for this.

**Comment by anpete** [comment]
The EF.Property method should be the preferred way of doing this.

**Comment by tdykstra** [comment]
@anpete Can you point me to sample code?

**Comment by anpete** [comment]
@tdykstra 

https://docs.microsoft.com/en-us/ef/core/modeling/shadow-properties
https://docs.microsoft.com/en-us/ef/core/api/microsoft.entityframeworkcore.ef

Although shadow properties are the primary use case, EF.Property can be used to address non-shadow props, too. We should probably update the links above to reflect that.

**Comment by tdykstra** [comment]
It looks like you have to hard-code the property type -- `Property<TProperty>(Object, String)`

What if you want to write code that uses a string to specify the property name but you don't know the type until runtime? As in my example, where I sort on either a DateTime or a string field.

**Comment by anpete** [comment]
You're right, but it makes it slightly easier. If you don't want to hard-code a known set of types, then something like this can work:
```c#
public static class DynamicLinqExtensions
{
    public static IQueryable<TElement> OrderBy<TElement>(
        this IQueryable<TElement> source,
        string propertyName,
        Type propertyType)
        => (IQueryable<TElement>)_dynamicOrderByMethodInfo
            .MakeGenericMethod(typeof(TElement), propertyType)
            .Invoke(null, new object[] { source, propertyName, false });

    public static IQueryable<TElement> OrderByDescending<TElement>(
        this IQueryable<TElement> source,
        string propertyName,
        Type propertyType)
        => (IQueryable<TElement>)_dynamicOrderByMethodInfo
            .MakeGenericMethod(typeof(TElement), propertyType)
            .Invoke(null, new object[] { source, propertyName, true });

    private static readonly MethodInfo _dynamicOrderByMethodInfo
        = typeof(DynamicLinqExtensions).GetTypeInfo().GetDeclaredMethod(nameof(DynamicOrderBy));

    private static IQueryable<TElement> DynamicOrderBy<TElement, TKey>(
        IQueryable<TElement> source,
        string propertyName,
        bool descending)
    {
        Expression<Func<TElement, TKey>> keySelector = e => EF.Property<TKey>(e, propertyName);

        return descending
            ? source.OrderByDescending(keySelector)
            : source.OrderBy(keySelector);
    }
}
```
which means you can write code like this in the application:
```c#
var customers = context.Customers.OrderBy("CompanyName", typeof(string));
```

**Comment by tdykstra** [comment]
> If you don't want to hard-code a known set of types

But you still have to hard-code the type in this example -- `OrderBy("CompanyName", typeof(string));`

**Comment by anpete** [comment]
That was just my example. You can replace that with `typeof(T).GetProperty(propertyName).PropertyType`

**Comment by divega** [comment]
@anpete do you think there could be something simple we should consider adding to EF Core to make specifying the type unnecessary? Or should we leave this to other libraries and "advanced" code like the extension method you showed? I just wonder if we could make an untyped version of `EF.Property()` work (even if it would have limitations).

**Comment by anpete** [comment]
@divega I just tested EF.Property with typeof(object) and it works :trollface: Not a general solution but makes this case (and likely projections also) easier.

**Comment by divega** [comment]
@tdykstra then if possible, I would change the sample to use `OrderBy(e => EF.Property<object>(e, "Property"))`

**Comment by tdykstra** [comment]
@bricelam Ready for re-review -- [new review URL](https://review.docs.microsoft.com/en-us/aspnet/core/data/ef-mvc/advanced?branch=td-dynamiclinq#use-dynamic-linq-to-simplify-sort-selection-code)

File: aspnetcore/data/ef-mvc/intro/samples/cu/Controllers/StudentsController.cs [hunk 1]
```diff
@@ -142,11 +142,70 @@ public StudentsController(SchoolContext context)
...
+            }
+
+            bool descending = false;
+            if (sortOrder.EndsWith("_desc"))
+            {
+                sortOrder = sortOrder.Substring(0, sortOrder.Length - 5);
+                descending = true;
+            }
+
+            if (descending)
```
Comment (divega) [inline aspnetcore/data/ef-mvc/intro/samples/cu/Controllers/StudentsController.cs]: @tdykstra the point I was trying to get across when I posted my previous comment is that you should be able to use `students.OrderBy(e => EF.Property<object>(e, propertyName))` directly without adding any new extension methods. Did you try that?

Comment (anpete) [inline aspnetcore/data/ef-mvc/intro/samples/cu/Controllers/StudentsController.cs]: Or modify the extension methods to not take the type as a parameter.

**Comment by tdykstra** [comment]
@divega - Sorry, missed that .. yes that code works.  Much simplified version is now in staging.

**Comment by tdykstra** [comment]
@anpete I opted not to modify the extension methods because this approach worked fine while requiring much less code.  If there are other benefits to using the extension methods I could add that as yet another option here.

File: aspnetcore/data/ef-mvc/intro/samples/cu/Controllers/StudentsController.cs [hunk 2]
```diff
@@ -142,11 +142,70 @@ public StudentsController(SchoolContext context)
...
+            else
+            {
+                searchString = currentFilter;
+            }
+
+            ViewData["CurrentFilter"] = searchString;
+
+            var students = from s in _context.Students
+                           select s;
+            if (!String.IsNullOrEmpty(searchString))
```
Comment (divega) [inline aspnetcore/data/ef-mvc/intro/samples/cu/Controllers/StudentsController.cs]: nit: CR before this?

**Pull request merged**